File: 6/second_min_node_in_btree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func secondMinNotEqualToVal(node *TreeNode, val int) int {
	if node.Left == nil {
		if node.Val == val {
			return -1
		} else {
			return node.Val
		}
	}
	secondMinLeft := secondMinNotEqualToVal(node.Left, val)
	secondMinRight := secondMinNotEqualToVal(node.Right, val)
	if secondMinLeft == -1 && secondMinRight == -1 {
		return -1
	} else if secondMinLeft == -1 {
		return secondMinRight
	} else if secondMinRight == -1 {
		return secondMinLeft
	} else {
		return min(secondMinLeft, secondMinRight)
	}
}

func findSecondMinimumValue(root *TreeNode) int {
	res := secondMinNotEqualToVal(root, root.Val)
	return res
}

func main() {
	root := TreeNode{2,
		&TreeNode{2, nil, nil},
		&TreeNode{5,
			&TreeNode{5, nil, nil},
			&TreeNode{7, nil, nil}}}

	res := findSecondMinimumValue(&root)
	fmt.Println(res)
}

File: 6/second_min_node_in_btree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
import sys

class Solution:
    def findSecondMinimumValue(self, root: TreeNode) -> int:
        def second_min(root, val):
            if not root.left:
                if root.val != val:
                    return root.val
                else:
                    return sys.maxsize
            else:
                return min(second_min(root.left, val), second_min(root.right, val))
        
        res = second_min(root, root.val)
        
        return res if res != sys.maxsize else -1
            
